process/smartContract/processorV2/processV2.go

```diff
@@ -0,0 +1,2786 @@
+		if ignorableError != nil {
+			log.Debug("scProcessor.finishSCExecution txLogsProcessor.SaveLog()", "error", ignorableError.Error())
+		}
+	}
+
+	return nil
+}
+
+func (sc *scProcessor) checkUpgradePermission(contract state.UserAccountHandler, vmInput *vmcommon.ContractCallInput) error {
+	isUpgradeCalled := vmInput.Function == upgradeFunctionName
+	if !isUpgradeCalled {
+		return nil
+	}
+	if check.IfNil(contract) {
+		return process.ErrUpgradeNotAllowed
+	}
+
+	codeMetadata := vmcommon.CodeMetadataFromBytes(contract.GetCodeMetadata())
+	isUpgradeable := codeMetadata.Upgradeable
+	callerAddress := vmInput.CallerAddr
+	ownerAddress := contract.GetOwnerAddress()
+	isCallerOwner := bytes.Equal(callerAddress, ownerAddress)
+
+	if isUpgradeable && isCallerOwner {
+		return nil
+	}
+
+	return process.ErrUpgradeNotAllowed
+}
+
+func (sc *scProcessor) createCompleteEventLogIfNoMoreAction(
+	tx data.TransactionHandler,
+	txHash []byte,
+	results []data.TransactionHandler,
+) *vmcommon.LogEntry {
+	sndShardID := sc.shardCoordinator.ComputeId(tx.GetSndAddr())
+	dstShardID := sc.shardCoordinator.ComputeId(tx.GetRcvAddr())
+	isCrossShardTxWithExecAtSender := sc.shardCoordinator.SelfId() == sndShardID && sndShardID != dstShardID
+	if isCrossShardTxWithExecAtSender && !sc.isInformativeTxHandler(tx) {
+		return nil
+	}
+
+	for _, scr := range results {
+		sndShardID = sc.shardCoordinator.ComputeId(scr.GetSndAddr())
+		dstShardID = sc.shardCoordinator.ComputeId(scr.GetRcvAddr())
+		isCrossShard := sndShardID != dstShardID
+		if isCrossShard && !sc.isInformativeTxHandler(scr) {
+			return nil
+		}
+	}
+
+	return createCompleteEventLog(tx, txHash)
+}
+
+func createCompleteEventLog(tx data.TransactionHandler, txHash []byte) *vmcommon.LogEntry {
+	prevTxHash := txHash
+	originalSCR, ok := tx.(*smartContractResult.SmartContractResult)
+	if ok {
+		prevTxHash = originalSCR.PrevTxHash
+	}
+
+	newLog := &vmcommon.LogEntry{
+		Identifier: []byte(completedTxEvent),
+		Address:    tx.GetRcvAddr(),
+		Topics:     [][]byte{prevTxHash},
+	}
+
+	return newLog
+}
+
+func isReturnOKTxHandler(
+	resultTx data.TransactionHandler,
+) bool {
+	return bytes.HasPrefix(resultTx.GetData(), []byte(returnOkData))
+}
+
+// this function should only be called for logging reasons, since it does not perform sanity checks
+func (sc *scProcessor) computeTxHashUnsafe(tx data.TransactionHandler) []byte {
+	txHash, _ := core.CalculateHash(sc.marshalizer, sc.hasher, tx)
+
+	return txHash
+}
+
+// IsPayable returns if address is payable, smart contract ca set to false
+func (sc *scProcessor) IsPayable(sndAddress []byte, recvAddress []byte) (bool, error) {
+	return sc.blockChainHook.IsPayable(sndAddress, recvAddress)
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (sc *scProcessor) IsInterfaceNil() bool {
+	return sc == nil
+}
+
+// GetTxLogsProcessor is a getter for txLogsProcessor
+func (sc *scProcessor) GetTxLogsProcessor() process.TransactionLogProcessor {
```

Comment:
debate-able as we use these only in tests. I will find a solution to these ones, will remain for now as they are

Reply:
will figure out a solution to this